Handle more cases in cfg_accessible

Urgau opened the pull request:
This PR tries to handle more cases in the cfg_accessible implementation by only emitting a "not sure" error only if we have partially resolved a path.

This PR also adds many tests for the "not sure" cases and for private items.

r? @petrochenkov

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6591196752?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/bc049c1ffc4cb2c78614f9f0f32da458170a2295/checks/6591196752/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
......................................................iii............................... 13200/13273
.........................................................................
failures:

---- [ui] src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs#edition2015 stdout ----

41    |                  ^^^^^^^
42 
43 error: not sure whether the path is accessible or not
43 error: not sure whether the path is accessible or not
-   --> $DIR/cfg_accessible-not_sure.rs:51:18
+   --> $DIR/cfg_accessible-not_sure.rs:52:18
45    |
46 LL | #[cfg_accessible(Trait::foo)]

48 
49 error: not sure whether the path is accessible or not
-   --> $DIR/cfg_accessible-not_sure.rs:58:18
-   --> $DIR/cfg_accessible-not_sure.rs:58:18
+   --> $DIR/cfg_accessible-not_sure.rs:59:18
51    |
52 LL | #[cfg_accessible(TypeAlias::A)]

54 
55 error: not sure whether the path is accessible or not
-   --> $DIR/cfg_accessible-not_sure.rs:67:18
-   --> $DIR/cfg_accessible-not_sure.rs:67:18
+   --> $DIR/cfg_accessible-not_sure.rs:68:18
57    |
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
58 LL | #[cfg_accessible(ForeignType::A)]

60 
61 error: not sure whether the path is accessible or not
-   --> $DIR/cfg_accessible-not_sure.rs:76:18
-   --> $DIR/cfg_accessible-not_sure.rs:76:18
+   --> $DIR/cfg_accessible-not_sure.rs:77:18
63    |
64 LL | #[cfg_accessible(AssocType::AssocType::A)]

66 
67 error: not sure whether the path is accessible or not
-   --> $DIR/cfg_accessible-not_sure.rs:81:18
-   --> $DIR/cfg_accessible-not_sure.rs:81:18
+   --> $DIR/cfg_accessible-not_sure.rs:82:18
69    |
70 LL | #[cfg_accessible(u8::A)]


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/conditional-compilation/cfg_accessible-not_sure.edition2015/cfg_accessible-not_sure.edition2015.stderr
To only update this specific test, also pass `--test-args conditional-compilation/cfg_accessible-not_sure.rs`

error in revision `edition2015`: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "edition2015" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/conditional-compilation/cfg_accessible-not_sure.edition2015" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2015" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/conditional-compilation/cfg_accessible-not_sure.edition2015/auxiliary"
stdout: none
--- stderr -------------------------------
error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:15:18
   |
LL | #[cfg_accessible(Struct::a)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:17:18
   |
   |
LL | #[cfg_accessible(Struct::b)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:19:18
   |
   |
LL | #[cfg_accessible(Struct::c)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:29:18
   |
   |
LL | #[cfg_accessible(Union::a)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:31:18
   |
   |
LL | #[cfg_accessible(Union::b)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:33:18
   |
   |
LL | #[cfg_accessible(Union::c)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:42:18
   |
   |
LL | #[cfg_accessible(Enum::B)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:52:18
   |
   |
LL | #[cfg_accessible(Trait::foo)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:59:18
   |
   |
LL | #[cfg_accessible(TypeAlias::A)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:68:18
   |
   |
LL | #[cfg_accessible(ForeignType::A)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:77:18
   |
   |
LL | #[cfg_accessible(AssocType::AssocType::A)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:82:18
   |
   |
LL | #[cfg_accessible(u8::A)] //~ ERROR not sure

error: aborting due to 12 previous errors
------------------------------------------



---- [ui] src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs#edition2021 stdout ----

41    |                  ^^^^^^^
42 
43 error: not sure whether the path is accessible or not
43 error: not sure whether the path is accessible or not
-   --> $DIR/cfg_accessible-not_sure.rs:51:18
+   --> $DIR/cfg_accessible-not_sure.rs:52:18
45    |
46 LL | #[cfg_accessible(Trait::foo)]

48 
49 error: not sure whether the path is accessible or not
-   --> $DIR/cfg_accessible-not_sure.rs:58:18
-   --> $DIR/cfg_accessible-not_sure.rs:58:18
+   --> $DIR/cfg_accessible-not_sure.rs:59:18
51    |
52 LL | #[cfg_accessible(TypeAlias::A)]

54 
55 error: not sure whether the path is accessible or not
-   --> $DIR/cfg_accessible-not_sure.rs:67:18
-   --> $DIR/cfg_accessible-not_sure.rs:67:18
+   --> $DIR/cfg_accessible-not_sure.rs:68:18
57    |
58 LL | #[cfg_accessible(ForeignType::A)]

60 
61 error: not sure whether the path is accessible or not
-   --> $DIR/cfg_accessible-not_sure.rs:76:18
-   --> $DIR/cfg_accessible-not_sure.rs:76:18
+   --> $DIR/cfg_accessible-not_sure.rs:77:18
63    |
64 LL | #[cfg_accessible(AssocType::AssocType::A)]

66 
67 error: not sure whether the path is accessible or not
-   --> $DIR/cfg_accessible-not_sure.rs:81:18
-   --> $DIR/cfg_accessible-not_sure.rs:81:18
+   --> $DIR/cfg_accessible-not_sure.rs:82:18
69    |
70 LL | #[cfg_accessible(u8::A)]


The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/conditional-compilation/cfg_accessible-not_sure.edition2021/cfg_accessible-not_sure.edition2021.stderr
To only update this specific test, also pass `--test-args conditional-compilation/cfg_accessible-not_sure.rs`

error in revision `edition2021`: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "edition2021" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/conditional-compilation/cfg_accessible-not_sure.edition2021" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2021" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/conditional-compilation/cfg_accessible-not_sure.edition2021/auxiliary"
stdout: none
--- stderr -------------------------------
error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:15:18
   |
LL | #[cfg_accessible(Struct::a)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:17:18
   |
   |
LL | #[cfg_accessible(Struct::b)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:19:18
   |
   |
LL | #[cfg_accessible(Struct::c)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:29:18
   |
   |
LL | #[cfg_accessible(Union::a)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:31:18
   |
   |
LL | #[cfg_accessible(Union::b)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:33:18
   |
   |
LL | #[cfg_accessible(Union::c)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:42:18
   |
   |
LL | #[cfg_accessible(Enum::B)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:52:18
   |
   |
LL | #[cfg_accessible(Trait::foo)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:59:18
   |
   |
LL | #[cfg_accessible(TypeAlias::A)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:68:18
   |
   |
LL | #[cfg_accessible(ForeignType::A)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:77:18
   |
   |
LL | #[cfg_accessible(AssocType::AssocType::A)] //~ ERROR not sure

error: not sure whether the path is accessible or not
  --> /checkout/src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:82:18
   |
   |
LL | #[cfg_accessible(u8::A)] //~ ERROR not sure

error: aborting due to 12 previous errors
------------------------------------------

```

</details>

petrochenkov commented on src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:
```diff
@@ -0,0 +1,85 @@
+
+struct Struct {
+    a: u8,
+    pub b: u8
```
Not much difference between pub and non-pub here, since the check in the same module.
It can either be removed as redundant or moved to a separate module to make difference.

petrochenkov commented on src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:
```diff
@@ -0,0 +1,85 @@
+    A(u8),
+}
+
+#[cfg_accessible(Enum::B)] //~ ERROR not sure
```
```suggestion
#[cfg_accessible(Enum::unresolved)] //~ ERROR not sure
```
Could you use something "obviously unresolved" for unresolved names instead of A/B/etc?

petrochenkov commented on src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:
```diff
@@ -0,0 +1,85 @@
+#[cfg_accessible(u8::A)] //~ ERROR not sure
+const A: bool = true;
+
+fn main() {}
```
Missing test cases:
- Variant::unresolved - error (for compatibility with variant types)
- Trait::unresolved (actually unresolved, not an existing method from `dyn Trait`)

petrochenkov replied:
Also, the same set of test cases, but with existing names, like
```rust
Struct::existing
Enum::ExistingVariant
...
```
, they should all produce "not sure" errors except for `Enum::ExistingVariant` and `Trait::existing_assoc_item` which produce true.

petrochenkov commented on src/test/ui/conditional-compilation/cfg_accessible.rs:
```diff
@@ -29,14 +29,6 @@ macro_rules! generate_accessible_expanded {
-
 fn main() {
     ExistingPublic;
     AccessibleExpanded;
```
Missing test cases:
- unresolved - false
- TraitAlias::unresolved - false

petrochenkov replied:
Also, the same set of test cases, but with existing names.

petrochenkov commented on compiler/rustc_resolve/src/macros.rs:
```diff
@@ -443,11 +443,16 @@ impl<'a> ResolverExpand for Resolver<'a> {
                 }
+                PathResult::NonModule(..) => {
+                    self.session
+                        .struct_span_err(span, "not sure whether the path is accessible or not")
```
Could you add a note saying that the type may potentially have associated items (that's why we are not sure).

petrochenkov replied:
And also a comment saying that if we get a partially resolved `NonModule` in one namespace, then we should get the same result in any other namespaces, so we can return early.

est31 commented on src/test/ui/conditional-compilation/cfg_accessible.rs:
```diff
@@ -8,12 +10,10 @@ mod m {
-// FIXME: Not implemented yet.
-#[cfg_accessible(m::ExistingPrivate)] //~ ERROR not sure whether the path is accessible or not
+#[cfg_accessible(m::ExistingPrivate)]
 struct ExistingPrivate;
```
It might be better to add something here that if the cfg evaluates to true, there is a compilation error.

Urgau commented:
I've address the review comments. This is ready for another review.

@rustbot ready

petrochenkov commented on src/test/ui/conditional-compilation/cfg_accessible-not_sure.rs:
```diff
@@ -0,0 +1,85 @@
+#[cfg_accessible(u8::is_ascii)] //~ ERROR not sure
+const B: bool = true;
+
+fn main() {}
```
Still missing test case:
- Enum::ExistingVariant - true

Urgau commented on src/test/ui/conditional-compilation/cfg_accessible-bugs.rs:
```diff
@@ -0,0 +1,24 @@
+#[cfg_accessible(Enum::Existing::existing)]
+const A: bool = true;
+
+// FIXME: Currently returns `false` but should be "not sure"
```
I have put a hack in https://github.com/rust-lang/rust/pull/97391/commits/a6156ca8588865f1c2b42220fac4411f8dfd2131 that seems to work but it's still a hack.

I'm pretty sure the problem lies [here](https://github.com/rust-lang/rust/blob/9fed13030c2a2ebd79bfb1fd8be4f768cbe8c9d9/compiler/rustc_resolve/src/ident.rs#L1510-L1517), I think it should return an `NonModule` instead of a `failure` like the rest of the conditions. I tried doing this but it's not trivial many diagnostics change sometimes dramatically (the error reported no longer make sense). See https://github.com/rust-lang/rust/commit/396ccf426934433debb0abc600dc43ce29b08214 for more information.

petrochenkov replied:
The `PathResult` and cases in which its variants are returned are pretty inconsistent.
It would be nice to refactor it to always keep partial resolutions, regardless of module-ness, but it indeed may affect some wider range of diagnostics (https://github.com/rust-lang/rust/commit/2ddd1bc4aeb8f50a8c42e2d1af17e819ebfb843e doesn't look like a correct or complete solution for that).

petrochenkov commented on src/test/ui/conditional-compilation/cfg_accessible-bugs.rs:
```diff
@@ -0,0 +1,19 @@
+// check-fail
+// known-bug
```
It was a really bad idea to allow skipping `//~ ERROR` annotations in `// known-bug` test.
If `//~ ERROR` annotations cannot be added together with `// known-bug`, then `// known-bug` is better removed here.

petrochenkov commented on compiler/rustc_resolve/src/macros.rs:
```diff
@@ -443,11 +443,22 @@ impl<'a> ResolverExpand for Resolver<'a> {
                 }
+                PathResult::NonModule(..) |
+                // HACK(Urgau): This shouldn't be necessary
+                PathResult::Failed { is_error_from_last_segment: false, .. } => {
```
This is kind of throwing the baby out with the bathwater.
But if you don't want to modify `fn resolve_path_with_ribs`, then it's the best that can be done, I guess.

Urgau replied:
Quite the contrary I want to modify it. Didn't you saw https://github.com/rust-lang/rust/pull/97391#discussion_r881976616 ?

Urgau commented:
I've addressed the review comments: removed `known-bug` and added the requested modification in `rustc_resolve`. I'm not fully sure about the last one and there are some diagnostics regression. But I want your opinion on it and it's ready for another review.

@rustbot ready

petrochenkov commented on src/test/ui/conditional-compilation/cfg_accessible-bugs.rs:
```diff
@@ -0,0 +1,18 @@
+trait TraitAlias = std::fmt::Debug + Send;
+
+// FIXME: Currently shows "cannot determine" but should be `false`
+#[cfg_accessible(unresolved)] //~ ERROR cannot determine
```
This is not a bug, if the compiler says "cannot determine" (rather than "not sure") then it's legitimately stuck, and it's unlikely we can do anything here.

Urgau replied:
The problem here is that we return an error instead of `false` (ie it is unresolvable).
The error message is irrelevant to the bug.

Urgau commented:
I've reverted the `failed -> NonModule` change as requested; I've posted the patch here https://gist.github.com/Urgau/7356282bc688ea3ab71b94412e86a2d1. I also responded to the last comment.

@rustbot ready

petrochenkov commented:
Thanks!
@bors r+

bors commented:
:pushpin: Commit b76d1125d4f1596e4c4481064fa6342990f1a985 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov b76d1125d4f1596e4c4481064fa6342990f1a985 -->
<!-- homu: {"type":"Approved","sha":"b76d1125d4f1596e4c4481064fa6342990f1a985","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit b76d1125d4f1596e4c4481064fa6342990f1a985 with merge 656eec8785e25a7fba568dba83190df2b0981daa...
<!-- homu: {"type":"BuildStarted","head_sha":"b76d1125d4f1596e4c4481064fa6342990f1a985","merge_sha":"656eec8785e25a7fba568dba83190df2b0981daa"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6743278448?check_suite_focus=true)
Approved by: petrochenkov
Pushing 656eec8785e25a7fba568dba83190df2b0981daa to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6743278448?check_suite_focus=true"},"merge_sha":"656eec8785e25a7fba568dba83190df2b0981daa"} -->

rust-timer commented:
Finished benchmarking commit (656eec8785e25a7fba568dba83190df2b0981daa): [comparison url](https://perf.rust-lang.org/compare.html?start=6dadfc06fe628d7a381a52b07714a7a849a6223d&end=656eec8785e25a7fba568dba83190df2b0981daa&stat=instructions:u).


### [Instruction count](https://perf.rust-lang.org/compare.html?start=6dadfc06fe628d7a381a52b07714a7a849a6223d&end=656eec8785e25a7fba568dba83190df2b0981daa&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6dadfc06fe628d7a381a52b07714a7a849a6223d&end=656eec8785e25a7fba568dba83190df2b0981daa&stat=max-rss)
<details>
<summary>Results</summary>


- Primary benchmarks: no relevant changes found
- Secondary benchmarks: mixed results

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | 3.3%     | 3.3% | 1         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -3.8%    | -4.4% | 2         |
| All 😿🎉 (primary) | N/A      | N/A | 0         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=6dadfc06fe628d7a381a52b07714a7a849a6223d&end=656eec8785e25a7fba568dba83190df2b0981daa&stat=cycles:u)
<details>
<summary>Results</summary>


- Primary benchmarks: 😿 relevant regressions found
- Secondary benchmarks: no relevant changes found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | 3.6%     | 3.8% | 2         |
| Regressions 😿 <br /> (secondary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | N/A      | N/A | 0         |
| All 😿🎉 (primary) | 3.6%     | 3.8% | 2         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
